library/core/src/char/methods.rs

```diff
@@ -1510,7 +1510,8 @@ impl char {
     #[rustc_const_stable(feature = "const_ascii_ctype_on_intrinsics", since = "1.47.0")]
     #[inline]
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, '0'..='9' | 'A'..='F' | 'a'..='f')
+        // Bitwise or can avoid need for branches in compiled code.
+        matches!(*self, '0'..='9') || matches!(*self as u32 | 0x20, 0x61..=0x66)
```

Comment:
To make this more maintainable, how about using `(b'a' as u32)..=(b'f' as u32)` instead?

Reply:
Unfortunately that would be a syntax error. I can't figure out a nice-looking way to use character literals here; `RangeInclusive::contains` isn't const, `('a' as u32 <= *self as u32 | 0x20) && ('f' as u32 >= *self as u32 | 0x20)` is very long.

Reply:
You can still use that range in a pattern if you define it as a `const` item first.

Reply:
One could make it slightly more concise:

```rust
let lower = *self as u32 | 0x20;
matches!(*self, '0'..='9') || (lower >= b'a' as u32 && lower <= b'f')
```

As an aside, there's a (currently private) `const ASCII_CASE_MASK: u8` in `core::num` which might be preferable to a literal `0x20`.

Reply:
> You can still use that range in a pattern if you define it as a `const` item first.

```
    const LOWER_ASCII: RangeInclusive<u32> = ('a' as u32)..=('f' as u32);
    matches!(*c, '0'..='9') || matches!(*c as u32 | 0x20, LOWER_ASCII)
```
[doesn't seem to compile](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d5c26475e124614b28e4bb4da5d40981)? I can get it to work with `A` and `F` const items though which I have done.